libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/utils/DirectionsResponseUtils.kt

```diff
@@ -53,39 +39,50 @@ suspend fun parseDirectionsResponse(
 suspend fun parseNativeDirectionsAlternative(
     dispatcher: CoroutineDispatcher,
     json: String,
-    options: RouteOptions?,
-): Expected<Throwable, DirectionsRoute> =
+    options: RouteOptions,
+): Expected<Throwable, NavigationRoute> =
     withContext(dispatcher) {
         return@withContext try {
+            // todo remove after https://github.com/mapbox/mapbox-navigation-native/issues/5142
+            // replacing all nulls that are used to pad the response array with fake routes
+            // because DirectionsResponse#routes cannot have null values
             val jsonObject = JSONObject(json)
-            val uuid: String? = if (jsonObject.has(UUID)) {
-                jsonObject.getString(UUID)
-            } else {
-                null
-            }
-
-            val jsonRoutes = jsonObject.getJSONArray("routes")
-            check(jsonRoutes.length() > 0) {
-                "route alternatives response should contain at least one route"
-            }
-            val jsonRoute: DirectionsRoute = jsonRoutes.let { routesArray ->
-                for (i in 0 until routesArray.length()) {
-                    if (!routesArray.isNull(i)) {
-                        // TODO simplify after https://github.com/mapbox/mapbox-java/issues/1292
-                        return@let DirectionsRoute.fromJson(
-                            routesArray.getJSONObject(i).toString(),
-                            options,
-                            uuid
-                        ).run {
-                            toBuilder()
-                                .routeIndex(i.toString())
-                                .build()
-                        }
+            val jsonRoutes = jsonObject.getJSONArray(ROUTES_RESPONSE_FIELD)
+            var alternativeRouteIndex: Int? = null
+            val nonNullRoutes = JSONArray()
+            for (i in 0 until jsonRoutes.length()) {
+                val nonNullRoute = if (!jsonRoutes.isNull(i)) {
+                    check(alternativeRouteIndex == null) {
+                        "alternative route object should have only one route"
                     }
+                    alternativeRouteIndex = i
+                    jsonRoutes[i]
+                } else {
+                    // build a fake route to fill up the index gap in the response
+                    JSONObject(
+                        DirectionsRoute.builder()
+                            .distance(0.0)
+                            .duration(0.0)
+                            .build()
+                            .toJson()
+                    )
                 }
```

Comment:
@LukasPaczos , sorry, don't understand this change.

As I see `parseNativeDirectionsAlternative` used to pick first not null route and build `DirectionsResponse` out of it. Now `parseNativeDirectionsAlternative` selects all routes replacing nulls by the empty `DirectionRoute`.
Do I understand the change correctly? Can you please help me understand the reason behind this change?

Reply:
That's right. This is an attempt to create the correctly sized collection of routes in `NavigationRoute#directionsResponse#routes`. If we didn't do that, the `NavigationRoute#directionsRoute` and `NavigationRoute#directionsResponse#routes(NavigationRoute#routeIndex)` would return different values which would be incorrect.

And the `NavigationRoute#routeIndex` has to be the exact value that was in the original response because it's needed for route refresh purposes.

Reply:
Why does `NavigationRoute#routeIndex` have to be exact?

Reply:
`NavigationRoute#routeIndex` is used for a refresh feature where we have to specify the UUID of the original response (`NavigationRoute#directionsResponse#uuid`), alongside the index of the route in the response (`NavigationRoute#routeIndex`), and the index of the leg we want to refresh.

Reply:
If the index wasn't correct, the refresh request would return refreshed annotations for a wrong route.